### PR TITLE
add support for K8s metrics API as a source of PVC usage data 

### PR DESCRIPTION
**Pull request by arturrez**
This is a PR to adds support for K8s metrics API as a source of PVC data conntecting to  kubelet via proxy API.
It's not enabled by default, but can be turned on via command line

### CODE REVIEW

**Comment by cupnes** [comment]
@arturrez Unfortunately, we cannot accept this PR. It's because this PR uses the kubelet Summary API and this API may be marked as deprecated in the future.[*1]

[*1] https://github.com/kubernetes/kubernetes/issues/68522

**Comment by arturrez** [comment]
k thank you. let me look into this. I still think that prometheus dependency is not needed as this info is available through k8s api

**Comment by arturrez** [comment]
@cupnes , I pushed a commit in which I switched to `/metrics` endpoint exposed by `kubelet` instead. Looking forward to your feedback

**Comment by daichimukai** [comment]
Thank you for your contribution, but I am not very positive about merging this PR. In my opinion, there are difficulties with getting metrics from kubelet instead of prometheus.

For example, I suspect that if there are multiple worker nodes, volumes mounted on pods on different nodes than the pvc-autoresizer controller will not be resized.

Also, I don't know if the pvc-autoresizer controller can connect to kubelet with its service account. Even if it were possible in our e2e test environments, I don't know if this is the case in all environments.

**Comment by arturrez** [comment]
no worries. Some k8s users don't use prometheus ( like us :) ) but are still interested in automatic pv resizing. Prometheus gather this data from kubelet so it seems to me that prometheus dependency can be removed.
regarding RBAC, `charts/pvc-autoresizer/templates/controller/clusterrole.yaml` in my PR addresses this issue allowing controller to access kubelet API

**Comment by daichimukai** [comment]
> Prometheus gather this data from kubelet 

I've never heard of this information. Could you provide me with documentation or other sources of information?

**Comment by arturrez** [comment]
> > Prometheus gather this data from kubelet
> 
> I've never heard of this information. Could you provide me with documentation or other sources of information?

I came across https://observability.thomasriley.co.uk/monitoring-kubernetes/metrics/kubelet-cadvisor/

**Comment by daichimukai** [comment]
@arturrez 
I After looking at the PR implementation for a bit, I understand that this PR enables us to scrape the metrics using the kubernetes api. It it acceptable for me.

Before we proceed the review of this PR, I have a few requests:

* Check the DCO.
* Rebase and organize your commits. It would be nice to have a commit message explaining what the commit is about.
* I feel `--use-kubelet` is ambiguous. How about `--use-k8s-metrics-api`?
* You are trying to make to use the kubernetes api the default, but please keep the current behavior that uses prometheus as the default.
* Please test both the mode using prometheus and the mode using kubernetes api with CI. Our another project has such settings ([ref](https://github.com/topolvm/topolvm/blob/995c3e2dc8e299fbdb8731f447893906b966a4e4/.github/workflows/e2e-k8s.yaml#L51-L65)). It may help.

**Comment by arturrez** [comment]
k thank you. will do.

**Comment by arturrez** [comment]
addressed feedback

File: runners/metrics_client.go [hunk 1]
```diff
@@ -34,6 +34,11 @@ func NewPrometheusClient(url string) (MetricsClient, error) {
 	}, nil
 }
 
+// NewKubeletClient returns a new kubelet client
+func NewKubeletClient() (MetricsClient, error) {
```
Comment (daichimukai) [inline runners/metrics_client.go:38]:
```suggestion
func NewKubernetesMetricsAPIClient() (MetricsClient, error) {
```

File: charts/pvc-autoresizer/values.yaml [hunk 2]
```diff
@@ -7,13 +7,17 @@ image:
...
-  pullPolicy:  # Always
+  pullPolicy: # Always
 
 controller:
   # controller.replicas -- Specify the number of replicas of the controller Pod.
   replicas: 1
 
   args:
+    useK8sMetricsApi: false
+    # controller.args.useK8sMetricsApii -- Use K8s metrics API to query volume stats.
```
Comment (daichimukai) [inline charts/pvc-autoresizer/values.yaml:18]:
```suggestion
    # controller.args.useK8sMetricsApi -- Use K8s metrics API to query volume stats.
```

**Review by daichimukai — CHANGES_REQUESTED** [review]
Please avoid merging the main branch to the topic branch; rebase the topic branch to the main branch instead.

**Comment by daichimukai** [comment]
> Please test both the mode using prometheus and the mode using kubernetes api with CI. Our another project has such settings (ref). It may help.

This should be addressed too.

**Comment by tropnikovvl** [comment]
Work continued in this https://github.com/topolvm/pvc-autoresizer/pull/242

**Pull request closed without merging**